### ISSUE TITLE
Standard MIME content-type

### ISSUE THREAD

**Issue by pavelnikolov** [wardi/jsonlines#19]
What do you think about adding new HTTP content-type for jsonlines data.
What about `application/jsonl`?

**Comment by jbaehr** [comment]
I'd rather prefer `application/json-lines` otherwise it may look like a typo ;-)

In addition to the Media Type, a registered structured suffix may be interesting. In my eyes even more useful, to create media types like `application/vnd.my-company.some-thing+json-lines`.

See also:
https://www.iana.org/assignments/media-types/
https://www.iana.org/assignments/media-type-structured-suffix/

@wardi have you considered filing a registration for a json-lines Media Type and structured suffix at IANA?

**Comment by karmakaze** [comment]
There is an IETF [RFC 7464](https://tools.ietf.org/html/rfc7464) for JSON Text Sequences that uses mime type: `application/json-seq`

It allows prefixing each JSON record with <RS> control character and requires ending each JSON record with <LF>.

**Comment by jbaehr** [comment]
This seems like a duplicate of #9. The whole purpose of the `Content-Type` header is to communicate the media type.

**Comment by whlavina** [comment]
The lack of a definitive  IANA Media Type for JSON Lines causes some difficulty for those of us using the format. In the interest of pushing the issue, I took the liberty of starting a conversation:
https://mailarchive.ietf.org/arch/msg/json/dWMWD0JDa2HiUYjWjLjrQExeIx4/

Perhaps someone here would like to join that thread?

Disclaimer: I am in no way affiliated with the IANA/IETF. I am merely interested in using the format, correctly.

**Comment by sp4ce** [comment]
@whlavina the response from Tim Bray was the most helpful and it looks nothing had happened since then. I'll copy the interesting bit here for reference

> to register a media type you need to link to a stable specification.  The contents of https://jsonlines.org/ probably don’t qualify, so the conventional thing would be to write an Internet-Draft which AFAICT would be the same as json-seq only without the leading "ASCII Record Separator (0x1E)" but retaining the trailing \n.

**Comment by frederikb** [comment]
Hi @sp4ce, good to see that someone is leading the way to an actual RFC!

I've noticied that AWS is (apparently) using JSON Lines for one of their products. I haven't seen a description of the actual output to know whether or not it is compatible with JSON Lines. In any case they are using the mime type `application/jsonlines`. Thoughts on `application/jsonl` vs. that one?

**Comment by tim-hitchins-ekkosense** [comment]
AWS Claim it's compatible with JSON Lines - it links to the JSON Lines homepage

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/S3DataExport.Output.html

**Comment by dwaite** [comment]
If there's still interest in doing this, I would recommend an informational track internet-draft (I-D) to describe the jsonlines specification, with an IANA considerations section registering the media type. The idea is that drafts work towards RFCs work towards standards on a long evolutionary track of internet draft to RFC, and potentially to being an internet standard.

IETF wants to deal with immutable and permanently available documents, so you will likely need represent the encoding and parsing requirements authoritatively within the I-D itself, using IETF nomenclature. There's a lot of references to this available, and [the JSON Text Sequences RFC](https://www.rfc-editor.org/rfc/rfc7464.html) is likely an excellent example.

I suspect there will be feedback that some areas are not needed. For example, your UTF-8 encoding rule does not have much left to it once you reference [the JSON RFC](https://www.rfc-editor.org/rfc/rfc8259.html). That RFC already mandates UTF-8 for everything other than closed ecosystems.At that point, you have to decide whether the application "advice" that they might want to escape the string to work on ASCII transports becomes something you might want to represent as an application note on the jsonlines site, and a discussion you  have with the IETF more broadly - after all, it would also affect JSON and json sequence data over such transports.

Conversely, you may want to be quite a bit more specific for the sake of interoperability, such as whether applications MUST be able to consume `\r\n` line separators, and what application behavior is mandated/desired if invalid JSON text (including things like lines of just whitespace) are encountered within a stream. Variance in behaviors have led to a lot of security issues - imagine if your security compliance or logging components stopped reading a JSON lines sequence at a newline, while your application logic ignored the blank line and kept going.

**Comment by dwaite** [comment]
The `x-` prefix on a subtype is intended only for private use, e.g. for types with no expectation of interoperability between implementations. In that sense, your `application/x-ndjson` may conflict with other people's `application/x-ndjson`, such as presence or absence of a leading `[` or of trailing `,`, or even someone deciding they might as well send it in Big5 rather than UTF-8.

The lack of an immutable standard (like a RFC with a number) means that ndjson three years from now may make changes along lines like these for robustness, but implementations do not have a clear way to explain what they are compatible with.

There are plenty of commercial products which use vendor and x-prefixed media types, and which do not attempt to define fixed/robust/interoperable behavior. It is a matter of what this project is going for, which is why my first words were "If there's still interest in doing this". 

In terms of ramifications, most SDOs (standard defining organizations) won't touch dependencies which do not have these and other formalisms, and may use things like publication in another SDO (like IETF) as a sign of that. That means ndjson/jsonlines may be used in public facing API, but a large category of interoperable standards work either wouldn't touch it, or will standardize their own similar effort.

**Comment by tim-hitchins-ekkosense** [comment]
> which will likely move to `application/ndjson` in the future when there's more adoption

Well that's the problem, it might happen, at some point in the future. Given the usage of JSON lines in various commercial products, we're suggesting we do that formalisation now - or at least start the process very soon!

**Comment by wardi** [comment]
I'd love to see this.

So do we copy-paste JSON-SEQ https://datatracker.ietf.org/doc/html/rfc7464 without the "ASCII Record Separator (0x1E)"? JSON-SEQ discusses detecting truncated records and continuing a fair bit, all of that could be removed in a new RFC.

> Conversely, you may want to be quite a bit more specific for the sake of interoperability, such as whether applications MUST be able to consume `\r\n` line separators, and what application behavior is mandated/desired if invalid JSON text (including things like lines of just whitespace) are encountered within a stream. Variance in behaviors have led to a lot of security issues - imagine if your security compliance or logging components stopped reading a JSON lines sequence at a newline, while your application logic ignored the blank line and kept going.

Rule 3 in https://jsonlines.org/ mentions that a compliant parser will be able to consume `\r\n` because `\r` is ignored as surrounding whitespace by a json parser. Doesn't hurt to repeat it though.

Lines of only whitespace are already invalid by rule 2 in https://jsonlines.org/ , but again it doesn't hurt to make this clear.

To be specific let's say that any line that doesn't parse as valid JSON should be treated as an invalid record *but* still counts as a record for the purpose of numbering the lines.

**Comment by GabenGar** [comment]
Should it count as a record? The whole point of something called _JSON Lines_ is that it stores lines of a well defined format called JSON, not arbitrary character sequences. Depending on the nature on malformed data in a line it might as well make all other lines after it invalid and blow up logs with parsing errors noise when the offender is a single line (a whole file).

**Comment by timtjtim** [comment]
> So do we copy-paste JSON-SEQ

I think RFCs are copyrighted so to copy paste you would need permission of the original author

**Comment by whlavina** [comment]
I'm glad to see continued discussion and forward movement. It's interesting to see that YAML just recently (this month) gained IANA media type registration... 22 years after the format was first created. If YAML can do it, JSON Lines can, too! If there's any need for help with the process, maybe we could ask the folks who pushed the YAML RFC?
* https://github.com/yaml/yaml-spec/issues/49
* RFC 9512

**Comment by tim-hitchins-ekkosense** [comment]
Here's the guidelines on how to write an Internet Draft

https://authors.ietf.org/en/home

**Comment by LuisLopez44431** [comment]
Yes need to go true me RFC---IETF text

**Comment by darrelmiller** [comment]
@whlavina  You folks are welcome to come join the HTTPAPI mailing list https://datatracker.ietf.org/wg/httpapi/about/ and we can chat about a path to registering this media type.  This is where the YAML media type registration RFC was created and we are working towards the OpenAPI one also.

There is ongoing discussion about allowing mediatype registrations to happen in the standards tree without necessarily going through the process of writing an RFC for the format.  https://www.ietf.org/archive/id/draft-ietf-mediaman-standards-tree-00.html  Although, this format might be simple enough that an RFC would straightforward.

**Comment by ferdnyc** [comment]
> There is ongoing discussion about allowing mediatype registrations to happen in the standards tree without necessarily going through the process of writing an RFC for the format. https://www.ietf.org/archive/id/draft-ietf-mediaman-standards-tree-00.html Although, this format might be simple enough that an RFC would straightforward.

As of last month, that (expired) draft is replaced by https://www.ietf.org/archive/id/draft-ietf-mediaman-standards-tree-01.html

**Comment by finom** [comment]
JSON lines is a perfect solution for JSON streaming that is a common task today for features that use LLMs. I attempted to use `application/jsonl` but, unfortunately, if I open an endpoint URL with the browser, the content of the endpoint is being downloaded instead of showing up as text (as `application/json` does). I've spent significant amount of time figuring that out, and for a while the best solution was to use `text/plain` content-type with a custom header, like `x-format=jsonlines`.  

I didn't enjoy this solution because I wanted the `content-type` represent the actual content type of the response and came up with an idea to use `text/plain` with a custom attribute `format`. 

```text
text/plain; format=jsonlines
```

At this case, the response is interpreted as text and isn't being downloaded, but also the actual type of the content can be read from the `content-type` header.

It looks kinda right and kinda wrong at the same time, but since we don't have any standard header yet, this can be used as a solid temporary solution. Let me know what you think.

**Comment by GabenGar** [comment]
`format` is not a valid `Content-Type` [parameter](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type#directives), simple as.

**Comment by finom** [comment]
@GabenGar I agree, but what would be a good alternative for self-described content type of JSON lines that can be read by any client that doesn't support the proposed one? Maybe the parameter should be prefixed with `x-`?

```
text/plain; x-format=jsonlines
```

**Comment by GabenGar** [comment]
For what purpose do you want to violate the http spec?

**Comment by finom** [comment]
> For what purpose do you want to violate the http spec?

Do I? Is it directly forbidden to use custom params?

**Comment by GabenGar** [comment]
You didn't answer the question.

**Comment by timtjtim** [comment]
> You didn't answer the question.

The question makes no sense. Clearly @finom doesn't want to violate any HTTP standards - which their suggestion doesn't.

See the actual specification, rather than just the developer reference: https://httpwg.org/specs/rfc9110.html#field.content-type

There's no exhaustive list of valid parameters, so there's no rule again using `format`.

**Comment by dwaite** [comment]
> JSON lines is a perfect solution for JSON streaming that is a common task today for features that use LLMs. I attempted to use `application/jsonl` but, unfortunately, if I open an endpoint URL with the browser, the content of the endpoint is being downloaded instead of showing up as text (as `application/json` does). I've spent significant amount of time figuring that out, and for a while the best solution was to use `text/plain` content-type with a custom header, like `x-format=jsonlines`.
> 
> I didn't enjoy this solution because I wanted the `content-type` represent the actual content type of the response and came up with an idea to use `text/plain` with a custom attribute `format`.
> 
> ```
> text/plain; format=jsonlines
> ```
> 
> At this case, the response is interpreted as text and isn't being downloaded, but also the actual type of the content can be read from the `content-type` header.
> 
> It looks kinda right and kinda wrong at the same time, but since we don't have any standard header yet, this can be used as a solid temporary solution. Let me know what you think.

You really should not use the text top level type unless your intention is that an end user (not developer) would be able to read the data without tooling, for this reason (you get end-user display of raw data by browsers when there is no system registered tooling)

I’d recommend if this is an API endpoint, instead leverage the Accepts header. If the http client does not prefer application/jsonlines over text/plain, they get a plaintext version with text/plain media type. You can then disable the text/plain accepts handler in production. 
 
> `format` is not a valid `Content-Type` [parameter](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type#directives), simple as.

Format is a valid media type parameter for text/plain (RFC 2646), but it dictates things like newline interpretation. Using it to say “no I really mean this other media type” is certainly not its intended use or how media types are intended to work. Having an X-format doesn’t really make that better - X is really meant for indicating a parameter or type may have conflicts in independent usage due to not being registered, not “I want to do something entirely different than the thing with this unprefixed name”. The latter is creating something knowingly confusing, even before you get to this idea being counter to the purpose of media types

**Comment by finom** [comment]
@dwaite I've just finished implementing your idea to switch between content-types based on `accepts` header and so far it works perfectly. Thank you for your comment, this solved a lot of time for me. The only note is that I needed to provide charset in order to make it look nice at the development tools of Chrome browser.


```
application/jsonl; charset=utf-8
```

If accepts doesn't include `application/jsonl`, the server returns `text/plain; charset=utf-8` so any client can access the data.

**Comment by obfischer** [comment]
Hi, is there still interesst in writing an RFC for a custom MIME type? I would like to support you. Always wanted the write an RFC ;-)

**Comment by macropay-solutions** [comment]
@obfischer for us, it would mean we can replace application/Json with application/jsonl when the response should be streames json lines with limit > 0 (meaning just a page).

**Comment by gwpl** [comment]
btw, do I understand correctly that output of claude code --output-format
json-stream... is actually JSONL and that's how they should allow to
specify it in a flag?

If yes, maybe can someone, more knowledgeable, file issue against them
pointing to issues here?

We may spres JSONL awareness through back linking?

**Comment by sp4ce** [comment]
@obfischer 

> Hi, is there still interest in writing an RFC for a custom MIME type? I would like to support you. Always wanted the write an RFC ;-)

I was really interested as well at the beginning, but I lack the experience and the understanding of all the required steps. It seems that it starts here https://authors.ietf.org/en/home and you can try to give it a shot, or we can have a call together on of these days to help us go through it. Let me know. Thanks.

**Comment by kytta** [comment]
> Hi, is there still interesst in writing an RFC for a custom MIME type? I would like to support you. Always wanted the write an RFC ;-)

@obfischer @sp4ce I'd be down to help too :) I'm a fan of the format and am quite irritated that this hasn't become a standard yet :D 

Having had a short look at the [Authors Resources](https://authors.ietf.org/), it does not seem _too_ complicated. I guess the best for us would be to create a repo from [this template](https://github.com/martinthomson/internet-draft-template) and collaborate there, GitHub-style, with Markdown editing and merge requests.

[As pointed out in the email thread](https://mailarchive.ietf.org/arch/msg/json/i4mJuEo-ZFELKpLJGJG-kkm3bjw/), we would need a stable spec to register a data type / syntax suffix, so let's think about that later and focus on the RFC first.

**Comment by obfischer** [comment]
Sorry, my last comment got lost. I am afraid I didn't hit the comment button before leaving. ;-)

Yes, let us simply start. I will provide a repository for our work today in the evening.

**Comment by sp4ce** [comment]
Why creating a new repo? We could use this no? I can create a new section in the menu called `RFC (draft)` and we can contribute there no?

**Comment by obfischer** [comment]
Ok, please do it ;-)

**Comment by kytta** [comment]
> Why creating a new repo? 

If we use the template, we get automatic lifting and rendering via the pipelines for free. But for the start, a draft-draft-RFC in this repo is enough :)

**Comment by obfischer** [comment]
Valid point. So let us create a new repository. 

What do you think?

**Comment by sp4ce** [comment]
I think I could plus the rendering into the jsonlines website if needed. For now I just create and empty section in the website.

Also @wardi we had project to migrate to jekyll #8 that would be good to do it, it would be event easier to plug the rendering then.

**Comment by ross-spencer** [comment]
Is there an update on this since June? Looking in the sp4ce repo, last update was also 4 months ago.

**Comment by brawer** [comment]
Curious, what would be needed to make progress here? JSON Lines is an extremely useful format, so it's unfortunate it still has no official, non-experimental [IANA media type](https://www.iana.org/assignments/media-types/media-types.xhtml). Glancing over this discussion, it looks like there's been much debate about what exact string to use. Of course this is a good discussion to have, but to be honest, I believe most people don't have strong opinions on such things — they just want to use _officially registered_ identifier, not something in the experimental namespace (which is by definition non-interoperable), and also not something stated on a random website. Thus, a big thank you to whoever goes ahead and files a registration request.

**Comment by devurandom** [comment]
It appears as if reality outpaced the discussion. e.g. https://learn.openapis.org/specification/media-types#supported-sequential-media-types lists "JSON Lines: application/jsonl" as a supported media type, but it is indeed still missing on the official IANA list https://www.iana.org/assignments/media-types/media-types.xhtml .

**Comment by marius-ciclistu** [comment]
It is time after 10 years since this thread was opened.

**Comment by kytta** [comment]
Hello everyone, let me give a quick update before everyone starts spamming the issue with "It's been so long, [any updates](https://any-updates.info/)?!"

I was the one who offered to help `@sp4ce` with a formal IETF draft. As it always is, life happened, and I didn't have any time to give away for this project :( I am slowly returning to it and will submit first draft hopefully soon™. Otherwise, there hasn't been anything going on in the repo.

Now, regarding the naming. **Personally**, I don't like `application/jsonl` and I agree with `@dzhus` that it might look like a typo; `application/json-lines` would be better. The file extension can be `.jsonl` IMO. The problem is, as `@devurandom` already said, is that **realistically** the proposal will reach a lot more people if we choose `application/jsonl`. People don't like waiting and have just picked it as a _de facto_ MIME type, without really taking IANA list in consideration. Picking `/jsonl` will close the gap instead of creating [yet another competing standard](https://xkcd.com/927/).

As such, I will be writing first sections of the draft with `application/jsonl` as MIME type and `.jsonl` as file extension. I will do my best to keep y'all up-to-date. Until then, please only leave helpful comments ;)

**Comment by aphedges** [comment]
In [Firefox 155 release notes for developers (Stable)](https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/155), Mozilla announced support for viewing JSON Lines files in Firefox:

> The JSON Viewer now opens JSON Lines (NDJSON) documents, which are served as `application/jsonlines`, `application/x-ndjson`, or `text/jsonl`, or have a `.jsonl` file extension.

I don't see `application/jsonlines` or `text/jsonl` mentioned in this issue yet, so I thought it's worth noting they have been implemented in a client.

**Comment by kytta** [comment]
> In [Firefox 155 release notes for developers (Stable)](https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/155), Mozilla announced support for viewing JSON Lines files in Firefox:
> 
> > The JSON Viewer now opens JSON Lines (NDJSON) documents, which are served as `application/jsonlines`, `application/x-ndjson`, or `text/jsonl`, or have a `.jsonl` file extension.

What a move by Mozilla to support unofficial MIME types I've never heard about 😂

Especially `text/jsonl` makes no sense whatsoever because every data serialization format, including JSON itself, uses `application/` 🤦‍♂️

**Comment by aphedges** [comment]
It turns out these release notes are missing the `application/jsonl` type that they also support.

The JSON Lines support was added in <https://github.com/mozilla-firefox/firefox/commit/a4e1189c40c6dce9afcc713d0017a9df632dcec3>, which includes the following comment:

```js
        // Same idea, but for JSON Lines (one JSON value per line):
        // application/jsonl (the standard type), the older
        // application/jsonlines and application/x-ndjson spellings, or a
        // .jsonl file served without a matching (or without any) content type.
```

The `text/jsonl` support was a separate request in [Show content-type: text/jsonl responses with the JSON Viewer](https://bugzilla.mozilla.org/show_bug.cgi?id=2060972) because apparently some software serves it that way. It was added in <https://github.com/mozilla-firefox/firefox/commit/a42cbce518dc>.

**Comment by brawer** [comment]
So, how's it going with the IANA registration request?

**Comment by kytta** [comment]
> So, how's it going with the IANA registration request?

@brawer Please read the whole thread. Registering a new MIME type requires a stable spec, which this repo nor ndjson doesn't have. The easiest and most correct way is to write and propose an RFC describing this format with a proper spec. See [my last comment](https://github.com/wardi/jsonlines/issues/19#issuecomment-5318672701) about the status.

TL;DR: it's in progress, but slow, because everyone has other things to do. Everybody is welcome to help :)

---

To everyone: Folks will post updates, when there are updates. Everything you see in this issue is the current state of things. There are no secret IANA submissions or side chats
